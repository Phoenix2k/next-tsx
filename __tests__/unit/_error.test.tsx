/// <reference types="jest"/>

import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { NextRouter } from 'next/router';
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import Renderer from 'react-test-renderer';
import ErrorPage from '../../src/pages/_error';

jest.mock('next/router', () => ({
  useRouter(): Partial<NextRouter> {
    return {
      route: '/404'
    };
  }
}));

describe('Error page', (): void => {
  let container: HTMLElement | null | undefined;

  const mockRouter: Partial<NextRouter> = {
    push: (url: string) => Promise.resolve(!!url)
  };

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  it('renders correctly', () => {
    act(() => {
      ReactDOM.render(
        // @ts-ignore: Test push method only
        <RouterContext.Provider value={mockRouter}>
          <ErrorPage />
        </RouterContext.Provider>,
        container
      );
    });
    expect(document.getElementById('error-code').innerHTML).toMatch('Unknown');
  });

  it('props work as expected', async () => {
    // @ts-ignore: Test status code only
    let props = await ErrorPage.getInitialProps({ res: { statusCode: 404 } });
    act(() => {
      ReactDOM.render(
        // @ts-ignore: Test push method only
        <RouterContext.Provider value={mockRouter}>
          <ErrorPage {...props} />
        </RouterContext.Provider>,
        container
      );
    });
    expect(document.getElementById('error-code').innerHTML).toMatch('404');

    // @ts-ignore: Test status code only
    props = await ErrorPage.getInitialProps();
    act(() => {
      ReactDOM.render(
        // @ts-ignore: Test push method only
        <RouterContext.Provider value={mockRouter}>
          <ErrorPage {...props} />
        </RouterContext.Provider>,
        container
      );
    });
    expect(document.getElementById('error-code').innerHTML).toMatch('Unknown');
  });

  it('matches snapshot', (): void => {
    const snapshotRender = Renderer.create(<ErrorPage />);
    const tree = snapshotRender.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
