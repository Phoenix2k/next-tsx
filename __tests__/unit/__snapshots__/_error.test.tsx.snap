// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error page matches snapshot 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100vh;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100vw;
}

.emotion-0 h1 {
  font-size: 2em;
  margin: 0;
  padding: 0;
}

.emotion-0 h2 {
  font-size: 6em;
  margin: 0.5rem;
}

<main
  className="emotion-0"
>
  <h1
    id="error-status-code"
  >
    Error
  </h1>
  <h2>
    <code
      id="error-code"
    >
      Unknown
    </code>
  </h2>
  <a
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    Return to home page
  </a>
</main>
`;
